_upgrading/sopel-8.md

```diff
@@ -0,0 +1,420 @@
+output can be skipped, retried after a delay, etc.
+
+_Hopefully no one was overwriting the old simple attribute in plugin code, but
+now they **can't** do that._
+
+  [sopel-sockmsg]: https://github.com/dgw/sopel-sockmsg
+
+### More consistent `trigger` objects
+
+#### `STATUSMSG` handling
+
+Events sent to a channel can be scoped to users with a particular privilege
+level, or _status_, in that channel. IRC servers advertise the availability of
+this feature using the `STATUSMSG` token in `RPL_ISUPPORT`.
+
+In Sopel 7, `trigger.sender` included the status prefix, and it was difficult
+for plugins to detect and remove it themselves if so desired. Sopel 8.0 removes
+the status prefix from `trigger.sender` and leaves only the channel name, while
+the status prefix (if any) is exposed as `trigger.status_prefix`. This change is
+useful for plugins that use `trigger.sender` to store or retrieve _channel
+values_ in the bot's database, for example.
+
+The `bot` passed to a plugin callable triggered by an IRC event automatically
+includes the status prefix, if present, in the default `destination` parameter
+to methods that send a _message_ to IRC. Plugin callables that simply invoke
+`bot.say()`, `bot.reply()`, etc. in direct response to a `trigger` and without
+overriding the default `destination` will most likely get the expected behavior.
+
+[`sopel-remind`][sopel-remind] is an example of a plugin that this change might
+surprise, since the naïve implementation of such a reminder plugin would simply
+store `(trigger.nick, trigger.sender, parsed_message)` and send
+`bot.reply(parsed_message, trigger.sender, trigger.nick)` after the specified
+amount of time has passed. The `trigger.status_prefix` is lost, and the reminder
+sent to the entire channel, rather than the status-limited subset of users who
+could have seen the original command.
+
+#### Only set the `sender` when it makes sense
+
+For general IRC events like `QUIT`, `RPL_NAMREPLY`, etc. that don't happen "in"
+a channel or other clearly defined _context_, the `trigger.sender` value that
+plugin callables saw in Sopel 7 was unpredictable at best.
+
+Sopel 8.0 now _only_ sets the `trigger.sender` if the triggering event warrants
+it. In all other cases, the `sender` will be `None` and plugin code should use
+the `trigger.args` list to retrieve all information about the event.
+
+#### To have `text`, you must first have `args`
+
+`trigger.text` will now be _empty_ (`''`) if the event carries no `args`.
+In Sopel 7, `trigger.text` contained the command name (e.g. `'QUIT'`) in these
+cases, but that was a bug. It has been fixed in Sopel 8.
+
+### IRC privilege requirement changes
+
+In Sopel 8.0, the `@sopel.plugin.require_privilege()` decorator now implies
+`@sopel.plugin.require_chanmsg()`, and the decorated callable will not run if
+triggered outside of a channel. Previously, `require_privilege` restrictions
+were simply ignored, and the callable would run anyway.
```

Comment:
Not really feedback on the migration guide, but when reading this originally, it struck me that I wished these decorators referred to "channel privilege" to better align with "bot privilege". Maybe something to consider for 9.0

Reply:
If this is something that hasn't been brought up yet (or at least recently), a better place to explore this would be the main issue tracker (see past issues with "RFC" in the title), or even better to start off in the IRC channel and flesh out the idea a little before writing it up as a planning issue.

`require_privilege` is [9 years old](https://github.com/sopel-irc/sopel/commit/3feb82ab24891fd5a4a6449f3e7e64cba4fd4865) and I wouldn't want to move too quickly toward removing it, but it could be converted to a compatibility alias pretty easily. (That's a structural topic for another day, too: Possibly making `sopel.plugin` into a full-fat submodule that lives in `sopel/plugin/__init__.py` so we can define compatibility shims in someplace like `sopel.plugin.compat`.)

Reply:
Now tracking as: https://github.com/sopel-irc/sopel/issues/2605